File: JavaScript/42. Trapping Rain Water.js
var trap = function(height) {
    let stack = [];
    let area = 0;

    // o(1) monotonic stack solution !
    // the code 

    for (let i = 0; i < height.length; i++) {
        while (stack.length && height[i] > height[stack.slice(-1)] ) {
            const top = stack.slice(-1);
            stack.pop();
            if (stack.length == 0) break;
            const h = Math.min(height[stack.slice(-1)], height[i]) - height[top];
            const d = i - stack.slice(-1) - 1;
            area += d * h;
        }
        stack.push(i);
    }
    return area;
}

File: Cooperative coding with my friends/monotonic stack pattern/42. Trapping Rain Water.py
class Solution:
    def trap(self, height: List[int]) -> int:

        # monotonic sort solution ---> descreasing stack!!

        stack = []
        res = 0

        for currentIndex in range(len(height)):
            while stack and height[stack[-1]] < height[currentIndex]:
                index = stack.pop()
                if not stack: break
                distance = currentIndex - stack[-1] -1
                value = min(height[currentIndex], height[stack[-1]]) - height[index]
                res += distance * value 
            stack.append(currentIndex)
        
        return res
    
# monotonic sort time complexity o(n)
# space complexity o(n)


# we can also use dynamic programming
class Solution:
    def trap(self, height: List[int]) -> int:
        if not height: return 0

        n = len(height)
        left_most = [0] * n
        right_most = [0] * n

        # initialization the boundary
        left_most[0] = height[0]
        right_most[-1] = height[-1]

        for i in range(1, n):
            left_most[i] = max(left_most[i-1], height[i])

        for i in range(n-2, -1, -1):
            right_most[i] = max(right_most[i-1], height[i])
        
        res = 0
        for i in range(n):
            val = min(left_most[i], right_most[i]) - height[i]
            if val >= 0:
                res += val
        
        return res

# time complexity o(n)
# space complexity o(n)

